Dec hwi int tdo

DobrowneyT opened the pull request:
This PR essentially adds the MCP23017 (IO Expander for solenoids) and the PCA9685 (PWM Generator for servo motors) into the ros2_control hardware system. Included with these two interface plug-ins is a universale I2CPeripheral that will be shared among the four I2C components in the robot (2x MCP23017 and 2x PCA9685). 

The drivers successfully launch (though it was only tested with a single "hand"), and the action server can pass commands to actuate the servo motors and solenoids. The main updates include:

- Singleton instance for I2C Peripheral as well as mutex locks to ensure no race conditions or errors.
- PCA9685 interface which includes the constants and methods for writing to the register and communicating with this device via I2C.
- MCP23017 interface which includes constant and methods for writing to the necessary registers to control the GPIO A pins. 

Not included is thorough documentation (sorry). These interfaces can also use some re-factoring and likely more deliberate error detection / handling. 

Also, @Sedkian was trying to modify the docker container so that it will have permissions to access the I2C bus, but it was not successful. To address this enter 
```sudo chmod 666 /dev/i2c-1``` 
into the terminal when you start-up the robot.

DobrowneyT committed:
Added pca9685_hardware_interface folder

DobrowneyT committed:
Added pca9685_hardware_interface to integrate PCA9685 support

DobrowneyT committed:
Moved the PCA9685 to src folder

DobrowneyT committed:
-Modified the PCA9685 package to take the i2c address from the .xacro and open the bus from there.
-Added the on_configure and on_cleanup methods

DobrowneyT committed:
-Modified the PCA9685 package to take the i2c address from the .xacro and open the bus from there.
-Added the on_configure and on_cleanup methods

DobrowneyT committed:
-Fixed spelling errors and other mistakes

DobrowneyT committed:
-Fixed spelling errors and other mistakes

DobrowneyT committed:
Re-structured folders in preparation for isolating the I2C code

DobrowneyT committed:
Added a common header include file and modified pca9685 files as necessary

DobrowneyT committed:
Updated includes filepaths and fixed error in CMakeLists

DobrowneyT committed:
Includes actually fixed now and builds

DobrowneyT committed:
Creation of I2CBusManager

DobrowneyT committed:
Creation of I2CBusManager

DobrowneyT committed:
Splitting plug-ins and re-defining I2CPeripheral

DobrowneyT committed:
Splitting plug-ins and re-defining I2CPeripheral

DobrowneyT committed:
Stashing multiple plug-in approach. See new branch hwi-tdo

DobrowneyT committed:
Stashing multiple plug-in approach. See new branch hwi-tdo

DobrowneyT committed:
-Updated the dockerfile to include libi2c-dev
-Made necessary changes to include the PC9685 in the .xacro and .xml file.
-Removed redundancies and cleaned up the folders

DobrowneyT committed:
-Modified the PCA repo to use positions instead of velocity
-Removed i2c_address in the initialization of the I2CPeripheral. Each component will connect to the peripheral with their unique address before sending any commands.
-Added the lib12c-dev to the package.xml

DobrowneyT committed:
-Moved the PCA constants to its respective folder, other cleaning up.

DobrowneyT committed:
-Added the MCP23017 interface

DobrowneyT committed:
-Fixed build errors

DobrowneyT committed:
Adjusted formatting

DobrowneyT committed:
Changes based on the ck-interfacing branch

DobrowneyT committed:
Creating init, setup, and connect functions in the mcp communication file

corbenkolaski committed:
Updated CMakeList linking

Sedkian committed:
Carrying Tristan: Temporary Fixes

corbenkolaski committed:
Corrected xacro property definitions.

Modified the stepper pulley radius to match the physical setup better, and corrected the previous carraige min and max position calculations.

DobrowneyT committed:
Just to get the latest updates

DobrowneyT committed:
-Re-organized the setp/init functions in the PCA and MCP system files to follow ros2_control lifecycle
-Updated the I2CPeripheral to create its own shared pointer and a function to return it. Updated MCP and PCA system files to get an instance of the shared pointer.
-Fixed some parameters information in the servo xacro to include duty cycles, and modified the PCA system file to reflect those parameters.

Sedkian committed:
namespace updates

Sedkian committed:
Removed unused files

Sedkian committed:
Added core to gitignore

Sedkian committed:
Fixed i2c permission errors

Sedkian committed:
changed bus name

Sedkian committed:
Removed Velocity from the command interfaces

Sedkian committed:
Removed old Bothoven Hardware

Sedkian committed:
Pedantic: Renamed parameters to include ms

Sedkian committed:
Latest system that will actually run.

Sedkian committed:
-Added the thread-safe singleton shared pointer for the I2C bus which appears to be working
-Added a clean_build.sh to ease with launching
-Made some modifications to the pca9685 so that it will launch. duty cycle is hard-coded at the moment, still need to make this dynamic
-Tristan made this commit, not Tony.

Sedkian committed:
-Re-named the clean/build/run script for clarit
-Modified MCP system file to only write on changes and added more debug messages

Sedkian committed:
-Added duty cycle as a parameter in the hardware interface (universal for all servos for now)
-Updated the logic MCP and PCA to receive commands and some error-handling
-Added an extra lock when opening the I2C bus
-Minor adjustments

Sedkian commented on src/hardware/include/hardware/i2c/I2CPeripheral.h:
```diff
@@ -0,0 +1,56 @@
+        static std::shared_ptr<I2CPeripheral> createInstance(const std::string& device);
+
+        // // Allow std::make_shared to access the private constructor
+        // friend std::shared_ptr<I2CPeripheral> std::make_shared<I2CPeripheral>(const std::string&);
```
What's up with the commented code? Do we need this?

corbenkolaski commented on scripts/clean_build_and_run.sh:
If this script specifically launches the hardware, the name should reflect that.

DobrowneyT replied:
Name was changed to reflect its function

corbenkolaski commented on src/hardware/src/pca9685/pca9685_system.cpp:
```diff
@@ -0,0 +1,206 @@
+        double offset = (max_duty_cycle + min_duty_cycle) / 2;
+
+        return slope * clamped_command + offset;
+    }
```
You should add comments to the `command_to_duty_cycle`  function to improve readability. However, if this is supposed to linearly scale values in [`min_input`, `max_input`] to values in [`min_duty_cycle`, `max_duty_cycle`] I have concerns about the implementation. I think `slope` is defined correctly, but shouldn't `offset` = `min_duty_cycle` and the returned value be `slope * (clamped_command - min_input) + offset`

DobrowneyT replied:
Added a description and changed the function to represent a linear conversion from one range to another.

corbenkolaski commented on src/hardware/src/i2c/I2CPeripheral.cpp:
The functionality looks good, but should we maintain consistent style conventions across the project (snake case)?

DobrowneyT replied:
I will leave this for now and wait for the re-factor branch.

corbenkolaski commented on src/hardware/src/pca9685/pca9685_system.cpp:
```diff
@@ -0,0 +1,206 @@
+            return hardware_interface::CallbackReturn::ERROR;
+        }
+
+        // Not sure if I need this second check
```
Is this check redundant?

corbenkolaski commented on src/hardware/src/pca9685/pca9685_system.cpp:
```diff
@@ -0,0 +1,206 @@
+    hardware_interface::return_type Pca9685SystemHardware::write(const rclcpp::Time & /*time*/,
+                                                                 const rclcpp::Duration & /*period*/) {
+        // Connect to the I2C bus (if this fails, no point in continuing)
+        if (!pca_.connect()) {
```
`connect` either returns 1 or will throw an exception. This if statement should be replaced with a try-catch or `connect` should be modified to catch the error and return an error code.

DobrowneyT replied:
This file now only passes the exceptiong, they will be caught in the system file and thrown from the I2CPeripheral class

corbenkolaski commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -0,0 +1,218 @@
+        // Only perform the write if the new state is different from the current state
+        if (new_solenoid_values_ != current_solenoid_values_) {
+            // Connect to the I2C bus (if this fails, no point in continuing)
+            if (!mcp_.connect()) {
```
`connect` either returns 1 or will throw an exception. This if statement should be replaced with a try-catch or `connect` should be modified to catch the error and return an error code.

corbenkolaski commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -0,0 +1,218 @@
+            return hardware_interface::CallbackReturn::ERROR;
+        }
+
+        // Not sure if I need this second check
```
Is this check redundant?

corbenkolaski commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -0,0 +1,218 @@
+                RCLCPP_ERROR(rclcpp::get_logger("Pca9685SystemHardware"), "Failed to connect to I2CBus during write.");
+                return hardware_interface::return_type::ERROR;
+            }
+            mcp_.set_gpio_state(new_solenoid_values_);
```
`set_gpio_state` calls `WriteRegisterByte` which can potentially throw an exception which wouldn't be caught

corbenkolaski commented on src/hardware/src/pca9685/pca9685_comm.cpp:
```diff
@@ -0,0 +1,95 @@
+        set_all_pwm(0, 0);
+        i2c_dev->WriteRegisterByte(MODE2, OUTDRV);
+        i2c_dev->WriteRegisterByte(MODE1, ALLCALL);
+        usleep(5'000);
```
The sleep duration should be a constant.

corbenkolaski commented on src/hardware/src/pca9685/pca9685_comm.cpp:
```diff
@@ -0,0 +1,95 @@
+        auto prescaleval = 2.5e7;  //    # 25MHz
+        prescaleval /= 4096.0;     //       # 12-bit
+        prescaleval /= freq_hz;
+        prescaleval -= 1.0;
```
These values should be constants. Also, could you add a comments to this method explaining:
1) what the prescale is, and why you subtract one here?
2) the purpose of the sequence of i2c writes

DobrowneyT replied:
Added comments and reference the datasheet formula

corbenkolaski commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -0,0 +1,218 @@
+        }
+
+        // Parse the MCP23017 parameters
+        cfg_.i2c_device = info_.hardware_parameters.at("i2c_device");
```
Parsing hardware parameters should occur in try-catch blocks since this will throw std::out_of_range if the parameter is not present.

corbenkolaski commented on src/hardware/src/pca9685/pca9685_system.cpp:
```diff
@@ -0,0 +1,206 @@
+        }
+
+        // Parse the MCP23017 parameters
+        cfg_.i2c_device = info_.hardware_parameters.at("i2c_device");
```
Parsing hardware parameters should occur in try-catch blocks since this will throw std::out_of_range if the parameter is not present.

corbenkolaski reviewed (commented):
Great work! You fixed the shared pointer issues I had with PR #8, and since we've tested this on one hand we have some confirmation that the implementation works. 

I still think we should confirm the hardware interfaces of each hand are run in a common process so that we have some certainty that this shared memory implementation will work, as otherwise that would require major changes.

I know we plan on making a new branch to refactor the codebase (add documentation, standardize organization and naming conventions) so some of my comments shouldn't prevent merging and can be handled later. However, I did have some concerns with exception handling, and the logic in a couple places which should be addressed.

Sedkian commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -0,0 +1,218 @@
+
+            for (int i = 0; i < 5; ++i) {
+                uint8_t bit_value = (current_solenoid_values_ >> i) & 0x01;
+                RCLCPP_INFO(
+                    rclcpp::get_logger("Mcp23017SystemHardware"),
+                    "Bit %d: %d", i, bit_value);
```
Pedantic: I think it would be better if we do not print in the write function itself as part of our efforts for a better real-time performance. Maybe move it out of the write function or change it from INFO to DEBUG.

DobrowneyT replied:
Sure. We discussed putting these outputs into DEBUG, or creating something like that. For now, this can just be commented out as necessary when running the system, it would not be something that it constantly printed when running the actual system.

Sedkian commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -0,0 +1,218 @@
+        const rclcpp_lifecycle::State & /*previous_state*/) {
+        RCLCPP_INFO(rclcpp::get_logger("Mcp23017SystemHardware"), "Configuring ...please wait...");
+
+        // try {
+        //     // Create or fetch the shared I2C bus instance
+        //     i2c_bus_ = std::make_shared<hardware::I2CPeripheral>(cfg_.i2c_device);
+
+        // } catch (const std::exception &e) {
+        //     RCLCPP_FATAL(rclcpp::get_logger("Mcp23017SystemHardware"), "Error initializing I2C Bus: %s", e.what());
+        //     return hardware_interface::CallbackReturn::ERROR;
+        // }
+
```
Get rid of the code that we are not using anymore.

Sedkian reviewed (approved):
Nicely done. I mostly reviewed your changes to the ROS2 architecture and linking and everything looks good to me. I haven't looked deeply into your interfacing code.

Sedkian commented:
> Also, @Sedkian was trying to modify the docker container so that it will have permissions to access the I2C bus, but it was not successful. To address this enter `sudo chmod 666 /dev/i2c-1` into the terminal when you start-up the robot.

Also, @DobrowneyT I thought it turned out to be working fine, no?

DobrowneyT commented:
> > Also, @Sedkian was trying to modify the docker container so that it will have permissions to access the I2C bus, but it was not successful. To address this enter `sudo chmod 666 /dev/i2c-1` into the terminal when you start-up the robot.
> 
> Also, @DobrowneyT I thought it turned out to be working fine, no?

I made this PR when I was not sure, we confirmed that it does work. However, you need to re-build the container when it loads up.

DobrowneyT committed:
Addressing comments from PR #11
- Improved error handling, try-catches should occur at the top system level
- Comments and clarification added in several sections
- Pedantic BS

Sedkian committed:
- Fixed some minor bugs and added error handling and the linear conversion function as per the comments [TDO]

corbenkolaski commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -194,21 +186,42 @@ namespace mcp23017_hardware_interface {
+                        RCLCPP_INFO(rclcpp::get_logger("Mcp23017SystemHardware"), "Bit %d: %d", i, bit_value);
+                    }
+                } catch (const std::exception &e) {
+                    if (num_write_attempts_ < MAX_WRITE_ATTEMPTS) {
```
num_write_attempts_ < MAX_WRITE_ATTEMPTS is included in the while loop condition so this will always be true. Remove the if-else from the catch block, and check num_write_attempts_ == MAX_WRITE_ATTEMPTS outside of the while loop

corbenkolaski commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -194,21 +186,42 @@ namespace mcp23017_hardware_interface {
+                    current_solenoid_values_ = new_solenoid_values_;
+                    write_success_ = true;
+
+                    for (int i = 0; i < 5; ++i) {
```
Use  hw_commands_.size() instead of hardcoding 5 loop iterations

corbenkolaski commented on src/hardware/src/pca9685/pca9685_comm.cpp:
```diff
@@ -1,94 +1,134 @@
+     * high. The off register is the point in the PWM cycle when the signal transitions from high to low. Therefore,
+     * each channel has 4 registers that need to be written to set the PWM signal.
+     * 
+     * @throw std::system_error if an error occurs during the I2C communication.
```
Also can throw std::out_of_range

corbenkolaski commented on src/hardware/src/mcp23017/mcp23017_system.cpp:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        // Validate position bounds
+        for (auto i = 0u; i < info_.joints.size(); i++) {
```
Why validate position bounds for the MCP but not the PCA?

corbenkolaski commented on src/hardware/src/pca9685/pca9685_system.cpp:
```diff
@@ -0,0 +1,224 @@
+                        write_success_ = true;
+                        current_command_values_[i] = hw_commands_[i];
+                    } catch (const std::exception &e) {
+                        if (num_write_attempts_ < MAX_WRITE_ATTEMPTS) {
```
num_write_attempts_ < MAX_WRITE_ATTEMPTS is included in the while loop condition so this will always be true. Remove the if-else from the catch block, and check num_write_attempts_ == MAX_WRITE_ATTEMPTS outside of the while loop

Sedkian committed:
Addressing Comments from PR #11
- Fixed the while loops for PCA and MCP, so they will output the re-try statements correctly.
- Validated position bounds in PCA
- Minor re-factoring and other cleanup/comments

corbenkolaski reviewed (approved):
The code looks good now. Once we finalize testing the functionality on the Pi this can be merged.
